### ISSUE TITLE
Discuss the example

### ISSUE THREAD

**Issue by wjwwood** [osrf/dynamic_config#2]
I wanted to point out and discuss some items with the current state of the repository, keep in mind I don't necessarily disagree with any of the choices I am just interested in your logic behind them.

First, I made some changes on top of your merged pull request, mostly whitespace, typos or style changes:

https://github.com/osrf/roscpp_param_api/compare/01f605a6b8a0b17471746dcdc3f9db0c26531cab...master

---

I'll start with the `parameter_example_node`. I would like to see if we can use inheritance with templating to allow functions like this one:

https://github.com/osrf/roscpp_param_api/blob/master/parameter_example/src/parameter_example_node.cpp#L12

To not have to be generically templated, but instead maybe only on the Parameter type (I don't want to have to template on Cache type necessarily).

This leads me to ask why you setup the Parameter class as inheriting from the Caching Policy as opposed to the other way around (using inheritance)?

https://github.com/osrf/roscpp_param_api/blob/master/parameter_example/include/ros/gsoc/parameter.h#L145

Another option would have been to wrap the CachePolicy selection and behavior up internally in the Parameter class, not using templating (on the Parameter class) or inheritance at all.

Again not saying these are better ways of doing it, I am just trying to get you to think about the API and how these decisions might change how it looks to the end user.

---

As you pointed out in the pull request your "Cached" parameter is only using the rosparameter server. What are you plans in the future, you mention that it should use the "subscribe" method. Can you talk about that a little bit?

---

Is the `polymorphic_map.h` header only used in the tests? If so do you plan to use it in the implementation? If yes then no are your answers, respectively, then maybe it should be moved to the tests folder.

---

I am going to open a pull request to change/extend the API which we can discuss next.

**Comment by abellagonzalo** [comment]
_This leads me to ask why you setup the Parameter class as inheriting from the Caching Policy as opposed to the other way around (using inheritance)?_
It is a technique I read in this book (http://en.wikipedia.org/wiki/Modern_C%2B%2B_Design) and I tried to implement it. But I realized I did it wrong and I've change the interface to a more ROS way. I'm looking how Services are implemented to do it similar.

Feel free to criticize the code, the design, my workflow and everything around the project.

---

To replicate the getCached() function I think we can reuse the new subscribe functionality. Internally, the object registers a callback to be notified when the param has a new value, the callback stores the value in a private member and when the user asks for it with the "get" method, the stored value is returned.

---

I've deleted the polymorphic_map because it is no longer used. I was going to use it to store the key/values in the Parameter Server, but not anymore.

---

I hope tomorrow I will have the subscribe functionality.
